relayer/relays/execution/main.go

```diff
@@ -50,34 +56,97 @@ func (r *Relay) Start(ctx context.Context, eg *errgroup.Group) error {
 		return err
 	}
 
-	listener := message.NewEthereumListener(
-		&r.config.Source,
-		ethconn,
+	headerCache, err := ethereum.NewHeaderBlockCache(
+		&ethereum.DefaultBlockLoader{Conn: ethconn},
 	)
-
-	err = listener.Start(ctx, eg)
 	if err != nil {
 		return err
 	}
 
-	addresses, err := r.config.Source.GetBasicChannelAddresses()
+	address := common.HexToAddress(r.config.Source.Contracts.OutboundQueue)
+	contract, err := contracts.NewOutboundQueue(address, ethconn.Client())
 	if err != nil {
 		return err
 	}
 
-	messages, err := message.New(
-		writer,
-		listener,
-		addresses,
-	)
+	opts := bind.WatchOpts{
+		Context: ctx,
+	}
+
+	messages := make(chan *contracts.OutboundQueueMessage)
+
+	key, err := types.EncodeToBytes(r.config.Source.LaneID)
 	if err != nil {
-		return err
+		return fmt.Errorf("encode to bytes: %w", err)
 	}
 
-	err = messages.Sync(ctx, eg)
+	sub, err := contract.WatchMessage(&opts, messages, [][]byte{key})
```

Comment:
So to confirm, `WatchMessage` will only subscribe to the recent message events and we still need some scanner logic to find messages with missing nonces?

Reply:
yeah, this is just a very simple, incomplete implementation that allows us to test the on-chain code. We will need to rewrite it in a future PR.

Reply:
Also we may have to modify the execution message relayer to support the optimizations we talked about on slack (execution headers, etc).

Reply:
Agree.  And not sure if `subscribe` will work here because we can not verify the execution block if it is not finized in beacon chain yet. so maybe the previous filter with `FilterOpts`(i.e. a bounded range) make more sense, the start of the range should be the `LatestVerifiedBlockNumber` in message layer(we remove it in this pr) and the end is the `LastestFinalized` in beacon layer?

Reply:
I check here that the execution header has already been imported: https://github.com/Snowfork/snowbridge/pull/791/files#diff-ba222152cf61a695305ad44e12ccc757ffa95a1f6c1c8ee698e21a7650155229R102

But yes, when we re-implement the relayer properly, we will likely use FilterMessage along with the following algorithm:
1. Check message nonce on Parachain side
2. Check message nonce on Ethereum side
3. Compare nonces to determine which messages need to be synced.
4. Scan Ethereum blocks using `FilterMessage` to find the `OutboundQueue.Message` event logs. This is fast, as Ethereum indexes events, so we can almost lookup event logs by nonce: https://github.com/Snowfork/snowbridge/blob/2bf31bfbef0d1d3092533668017c3ad31aa8c680/core/packages/contracts/src/IOutboundQueue.sol#L7

Then as you say, we will use `LatestExecutionHeaderState` to check that execution header has been imported first.